README.md
# Python API Code Generator
Generates a $language module out of the existing API docs webpage. 
Model is that API docs page is the source of truth, and that libraries 
should depend upon that.  

Target APIs are currently Meraki and Trello.

## Problem
The Meraki API is constantly adding API calls. Projects exist in multiple
programming languages to access the API but do not get updated. Meraki-Apigen
generates code in your $language with the most recent API docs so that every
supported language has the most recent API call accessors available.

## Description
The module generated will have methods for any new API calls.
For this reason, it may make sense to rerun merakygengen occasionally
to get an up-to-date script in your $language.

This is still under active development.

## Features
### CLI options
#### --classy (python only)
Aggregate functions into classes based on their API section. 

#### --lint (python only)
Use the linting utility for $language to verify code quality for 
generated code.

* Python: pylint
* Ruby: rubocop

#### --textwrap
Wrap text according to the style guide for $language.

*Max Line Length*
* Python: 79
* Ruby: 120
* Powershell: 100

*Indentation*
* Python: 4 spaces
* Ruby: 2 spaces
* Powershell: 4 spaces

#### --sample-resp
Add the sample response to the function docstring.

### Languages
**Supported**
* python
* ruby

**In Progress**
* powershell

**Planned**
* go
* node

## Usage
### Powershell
import-module merakygen

## Similar Projects
Check out https://github.com/CiscoDevNet/awesome-merakiapis.
Most projects have functionality, but some have been abandoned for years.

CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]
* Change output of make_powershell_module from one file to a folder of 
functions (most Powershell projects use this project structure style.)
* Powershell encodes url params
* Apidocs now gets cached to reduce network requests (last modified needs
  to be < 24 hours)

## [0.2.1] - 2019-02-04
### Added
* Added this Changelog

## [0.2.0] - 2019-01-31
### Added
* Created toy ruby script to verify functionality
* Added some content to make_ruby_script as well based upon it.
* Ruby generated script works

### Changed
* Refactored Main function: Added line for cli options in preamble
* This will allow users to see what options were used to generate this
    specific file.

## [0.1.1] - 2019-01-30
### Added
* Add --add-sample-resp option: Added the ability to add the sample 
    response from the API docs to the function. 
    This can help with understanding what will be returned.
        
## [0.1.0] - 2019-01-30
### Added
* Added support for classes, pylint, func descriptions
* Python script working
* Initial tests have also been written.

## [0.0.0] - 2019-01-27
### Added
* Project Structure

<!---
CHANGELOG TYPES

Added:      for new features. 
Changed:    for changes in existing functionality.
Deprecated: for soon-to-be removed features.
Removed:    for now removed features.
Fixed:      for any bug fixes.
Security:   for vulnerability fixes.
-->

[Unreleased]: https://github.com/pocc/merakygen/compare/v0.2.1...HEAD
[0.2.0]: https://github.com/pocc/merakygen/compare/v0.2.0...v0.2.1
[0.2.0]: https://github.com/pocc/merakygen/compare/v0.1.1...v0.2.0
[0.1.1]: https://github.com/pocc/merakygen/compare/v0.1.0...v0.1.1
[0.1.0]: https://github.com/pocc/merakygen/compare/v0.0.0...v0.1.0
[0.0.0]: https://github.com/pocc/merakygen/commit/36e4cefbafadba3999870bcbcd46bdd8fc7ca351

merakygen/_cli.py
# -*- coding: utf-8 -*-
# Copyright 2019 Ross Jacobs All Rights Reserved.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
"""Merakygen:
    Code generator for the Meraki API

USAGE:
    merakygen (--key <apikey>) [--language <name>] [--targetapi <api>]
                  [--classy] [--lint] [--textwrap] [--sample-resp]
                  [-h | --help] [-v | --version]

DESCRIPTION:
    Convert all of the recently released Meraki v0 API calls into
    Python or Ruby. As new API calls are released all the time, rerun
    this occasionally.

OPTIONS:
  --key <apikey>        Your API key. You can find it by going to your profile.
  --language <name>     Create a script in language. Valid options are
                        python, ruby, bash, and powershell. If no language is
                        specified, python will be used.
                        For ruby linting, ruby/gem will need to be installed.
  --targetapi <api>     The API that is being targeted. Default is Meraki.
  -c, --classy          Use classes instead of a function list.
  -l, --lint            Call Pylint. If not 10.00/10, print error text.
  -r, --sample-resp     Add the sample response to function documentation.
  -t, --textwrap        Wrap text according to language. Python(79), Ruby(120)
                        Default is to wrap.
  -h, --help            Print this help message.
  -v, --version         Print version and exit.

SEE ALSO:
    Contact: Ross Jacobs (rosjacob [AT] cisco.com)
    Github: https://github.com/pocc/merakygen
"""
import sys
import subprocess as sp
import re

import docopt

from requests import __version__ as requests_version
from yapf import __version__ as yapf_version
from merakygen import __version__ as apigen_version


def get_bash_version():
    """Get the bash version if it exists and a message if it does not.

    Should return a string like '3.2.57(1)-release (x86_64-apple-darwin18)'
    """
    try:
        cmd_list = ['bash', '--version']
        sp_pipe = sp.Popen(cmd_list, stdout=sp.PIPE)
        sp_stdout = sp_pipe.communicate()[0].decode('utf-8')
        version_info = sp_stdout.split('version ')[1].split('\n')[0]
        return version_info
    except FileNotFoundError:
        msg = 'not found (required for bash testing)'
        return msg


def get_ruby_versions():
    """Check whether ruby/gem is installed. If so, then rubocop can be used."""
    try:
        sp_ruby_ver = sp.Popen(['ruby', '-v'], stdout=sp.PIPE)
        sp_gem_ver = sp.Popen(['gem', '-v'], stdout=sp.PIPE)
        ruby_ver = sp_ruby_ver.communicate()[0].decode('utf-8')
        gem_ver = sp_gem_ver.communicate()[0].decode('utf-8')
        return re.sub(r'\n|ruby ', '', ruby_ver), re.sub(r'\n', '', gem_ver)
    # If not installed, catch the error and return not found.
    except FileNotFoundError:
        msg = 'not found (required for ruby linting and testing)'
        return msg, msg


def get_powershell_version():
    """Get the powershell version if it exists and a message if it does not.

    Returns a string like 'PowerShell 6.1.2'
    """
    try:
        cmd_list = ['pwsh', '--version']
        sp_pipe = sp.Popen(cmd_list, stdout=sp.PIPE)
        version_info = sp_pipe.communicate()[0].decode('utf-8')
        return version_info
    except FileNotFoundError:
        msg = 'not found'
        return msg


def show_cli():
    """Show the docopt cli"""
    args = docopt.docopt(__doc__)
    if args['--version']:
        python_ver = sys.version.replace('\n', '')
        print('Meraki-APIgen', apigen_version, '\n\nPython', python_ver)
        print('\trequests', requests_version, '\n\tyapf', yapf_version)
        print('Testing/linting')
        ruby_ver, gem_ver = get_ruby_versions()
        print('\truby', ruby_ver, '\n\tgem', gem_ver)
        print('\tbash', get_powershell_version())
        print('\tbash', get_bash_version())
        sys.exit()
    # Python is default
    if not args['--language']:
        args['--language'] = 'python'
    if args['--language'] not in ['bash', 'python', 'ruby', 'powershell']:
        raise ValueError("Only valid languages are python, ruby, "
                         "bash, and powershell.\n" + __doc__)

    # These are all of the user-selected options.
    required_args = ['--key', '--language']
    options = [arg for arg in args if args[arg] and arg not in required_args]
    return args['--key'], args['--language'], options
